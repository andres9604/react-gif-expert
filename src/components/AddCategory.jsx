import React, { useState } from "react";
import PropTypes from "prop-types";

//componente especializado para el input

//el componente recibe una funcion llamada onNewCategory, que ees la que se encargara de almacenar la nueva categoria
export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  //expresion de funcion cuando cambia el input
  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  //expresion de funcion, cuando se envia el formulario
  const handleSubmitForm = (e) => {
    console.log("ejecucion formulario ...");

    //cancelamos el envio del formulario
    e.preventDefault();

    const cleanInputValue = inputValue.trim();

    //si el valor ingresado es menos de 1 caracter, no continua con el proceso
    if (cleanInputValue.length <= 1) return;

    //* PRIMERA OPCION al hook del padre GifExpert llamado setCategories, le pasamos el valor ingresado y enviado en el formulario

    //se maneja el hook como un callback, para asi poder acceder a los valores que ya traia y agregarle el nuevo valor
    //setCategories((categories) => [inputValue, ...categories]);

    //* SEGUNDA OPCION: le paso el valor a la funcion del componente padre, y es desde el padre donde se agrega la nueva categoria
    onNewCategory(cleanInputValue);

    //limpiamos el input
    setInputValue("");
  };

  //no es necesario usar un fragmento <> ya que el form es el elemento padre y encapsula todos los inputs
  return (
    <form onSubmit={handleSubmitForm} aria-label="form">
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Buscar gifs"
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
