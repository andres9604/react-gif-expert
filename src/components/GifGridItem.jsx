//se improta propTypes para validacion de tipado de las propiedades del componete
import PropTypes from "prop-types";

export const GifGridItem = ({ image }) => {
  return (
    <div className="card">
      <img src={image.url} alt={image.title} />
      <p>{image.title}</p>
    </div>
  );
};

//proptypes del componente
GifGridItem.propTypes = {
  image: PropTypes.object.isRequired,
};
