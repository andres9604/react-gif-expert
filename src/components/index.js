//Centralizamos los hooks de componentes

//exportamos todo lo que exporte cada uno de los componentes, asi, cuando se requiera alguno  deellos, todos se pueden llamar de la carpeta components, como se ve en componente GifExpertAPp
export * from "./AddCategory";
export * from "./GifGrid";
export * from "./GifGridItem";

// //importamos todos los componentes
// import { AddCategory } from "./AddCategory";
// import { GifGrid } from "./GifGrid";
// import { GifGridItem } from "./GifGridItem";

// export { AddCategory, GifGrid, GifGridItem };
