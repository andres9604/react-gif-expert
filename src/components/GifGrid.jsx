import { GifGridItem } from "./GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";

//Componente GifGird
export const GifGrid = ({ category }) => {
  //! NO ES BUENA PRACTICA DEFINIR LAS FUNCIONES QUE HACEN REQUEST DENTRO DEL COMPONENTE, PORQUE EN CADA RENDER DEL COMPONENTE SE REALIZARA EL REQUEST

  // ! NUNCA SE DEBE PONER EL LLAMADO DE UNA FUNCION DENTRO DE UN COMPONENTE
  //getGifs(category);

  /**********************************************/
  //PRIMERA OPCION PARA  EL COMPONENTE

  // //se crea el hook de images
  // const [images, setImages] = useState([]);

  // //* como el getGifs es una promesa, y debo esperar a que finalice, debo llamarla dentro de otra funcion, porque el await no es permitido dentro del hook useEffect

  // //funcion que consultara las imagenes y las asignara al stateSetImages
  // const getImages = async () => {
  //   //obtiene las imagenes de la API
  //   const imagesList = await getGifs(category);

  //   //asigna las imagenes al hook setImages (para que las pinte en el componente)
  //   setImages(imagesList);
  // };

  // //para ejecutar funciones dentro de un componente, se utiliza el hook useEffect
  // useEffect(() => {
  //   //llamo la funcion que me ejecuta la consulta en la api
  //   getImages();
  // }, []);

  /**********************************************/
  //SE PASA TODO LO DE ARRIBA A UN CUSTOM HOOK LLAMADO useFetchGifs
  const { images, isLoading } = useFetchGifs(category);
  //console.log(images, isLoading);

  return (
    <>
      <h3>{category}</h3>

      {/* AND Logico */}
      {isLoading && <h2>Cargando ...</h2>}

      <div className="card-grid">
        {/* recorro el listado de las imagenes obtenidas de la api */}
        {images.map((image) => (
          <GifGridItem key={image.id} image={image} />

          // {...image} -> de esta forma enviara todas las properties de image al componente gifgriditem, y alla se recibira las props que hacen parte de image, id, title y url
        ))}
      </div>
    </>
  );
};
