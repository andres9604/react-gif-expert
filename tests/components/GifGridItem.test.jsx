/** @jest-environment jsdom */
import { render, screen } from "@testing-library/react";

import { GifGridItem } from "../../src/components/GifGridItem";

describe("pruebas en componente <GifGridItem />", () => {
  //objeto para probar el componente
  const objImage = {
    title: "saitama sensei",
    url: "https://saitama.com/sensei",
  };

  /******************************/

  test("Debe hacer match con el snapshot", () => {
    //renderizamos el componente
    const { container } = render(<GifGridItem image={objImage} />);

    //realizamos la validacion
    expect(container).toMatchSnapshot();
  });

  /******************************/

  test("Debe de mostrar la imagen con el URL  yel ALT indicado", () => {
    //renderizamos el componente
    const { container } = render(<GifGridItem image={objImage} />);

    // //debug el componente
    // screen.debug();

    //destructuring de los atributos de img
    const { src, alt } = screen.getByRole("img");

    //evaluamos que el src de la imagen sea igual a la url pasada en el componete
    expect(src).toBe(objImage.url);

    //evalua que el alt de la iamgen sea igual al title
    expect(alt).toBe(objImage.title);
  });

  /******************************/

  test("Debe de mostrar el titulo en el componente", () => {
    //renderizamos el componente
    render(<GifGridItem image={objImage} />);
    expect(screen.getByText(objImage.title)).toBeTruthy();
  });
});
