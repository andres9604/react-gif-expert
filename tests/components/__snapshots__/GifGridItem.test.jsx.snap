// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en componente <GifGridItem /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="saitama sensei"
      src="https://saitama.com/sensei"
    />
    <p>
      saitama sensei
    </p>
  </div>
</div>
`;
